Adding intel optimizations to latest torch and tf environment image builds

kta-intel opened the pull request:
Accidentally closed PR #665 after syncing fork with main branch in attempt to resolve some merge conflicts. The merge conflicts should now be resolved and I have added back in the original changes from the previous PR in the new directory structure

@savitamittal1 @elesel
Can you please re-approve and merge (if no additional conflicts). Thank you

@microsoft-github-policy-service agree company="Intel Corporation"

elesel commented:
@kta-intel would you please merge the latest main branch into yours? I think you're hitting some issues in the automated checks even though you haven't touched some of these files.

kta-intel commented:
> @kta-intel would you please merge the latest main branch into yours? I think you're hitting some issues in the automated checks even though you haven't touched some of these files.

I'm not sure why the pytorch env test is failing. You're right in that it looks to be triggered by a file that I haven't touched.
Should I try creating an entirely new fork and opening up another PR or did you have any recommendations?

elesel commented:
> > @kta-intel would you please merge the latest main branch into yours? I think you're hitting some issues in the automated checks even though you haven't touched some of these files.
> 
> I'm not sure why the pytorch env test is failing. You're right in that it looks to be triggered by a file that I haven't touched. Should I try creating an entirely new fork and opening up another PR or did you have any recommendations?

@kta-intel Merging the latest main branch did help, and the correct set of changed files is being used to determine which environments to build, based on changes to their directory contents. The new test you added is failing, though, and the root cause isn't coming through via the error handling at the end of the test. I've attached the contents of user_logs/std_log_process_0.txt here, which I hope helps you troubleshoot:
[std_log_process_0.txt](https://github.com/Azure/azureml-assets/files/11982780/std_log_process_0.txt)

kta-intel commented:
> @kta-intel Merging the latest main branch did help, and the correct set of changed files is being used to determine which environments to build, based on changes to their directory contents. The new test you added is failing, though, and the root cause isn't coming through via the error handling at the end of the test. I've attached the contents of user_logs/std_log_process_0.txt here, which I hope helps you troubleshoot: [std_log_process_0.txt](https://github.com/Azure/azureml-assets/files/11982780/std_log_process_0.txt)

Thanks @elesel. the error log is very helpful. 

The error looks to be coming from [pretrain_glue.py](https://github.com/Azure/azureml-assets/blob/main/assets/training/general/environments/acpt-tests/src/pretrain_glue.py) which is not one of our added tests. However, it does seem to be caused by how `DeepSpeed` is handling `intel_extension_for_pytorch`. Because no accelerator is set in [pytorch2_0_sample_test.py](https://github.com/Azure/azureml-assets/blob/main/assets/training/general/environments/acpt-pytorch-2.0-cuda11.7/tests/pytorch2_0_sample_test.py) `DeepSpeed` by default will try to use CPU acceleration before cuda (See: [real_accelerator.py](https://github.com/microsoft/DeepSpeed/blob/d24629f4fdaaa92df068de24f926d341f129112c/accelerator/real_accelerator.py#L80)). This means that it will try to use `intel_extension_for_pytorch`, which requires `oneCCL` to be compatible with `DeepSpeed`.

I could potentially include `oneCCL` as part of the environment, but I think the best path for now is to set the accelerator to cuda in the test script (which I will go ahead a do shortly). We'll leave `oneCLL` installation for a user to manually do since it isn't a requirement for most `intel_extension_for_pytorch` optimizations and then open a discussion for how/if we want to include it prebuilt in the environment itself.

elesel commented:
> > @kta-intel Merging the latest main branch did help, and the correct set of changed files is being used to determine which environments to build, based on changes to their directory contents. The new test you added is failing, though, and the root cause isn't coming through via the error handling at the end of the test. I've attached the contents of user_logs/std_log_process_0.txt here, which I hope helps you troubleshoot: [std_log_process_0.txt](https://github.com/Azure/azureml-assets/files/11982780/std_log_process_0.txt)
> 
> Thanks @elesel. the error log is very helpful.
> 
> The error looks to be coming from [pretrain_glue.py](https://github.com/Azure/azureml-assets/blob/main/assets/training/general/environments/acpt-tests/src/pretrain_glue.py) which is not one of our added tests. However, it does seem to be caused by how `DeepSpeed` is handling `intel_extension_for_pytorch`. Because no accelerator is set in [pytorch2_0_sample_test.py](https://github.com/Azure/azureml-assets/blob/main/assets/training/general/environments/acpt-pytorch-2.0-cuda11.7/tests/pytorch2_0_sample_test.py) `DeepSpeed` by default will try to use CPU acceleration before cuda (See: [real_accelerator.py](https://github.com/microsoft/DeepSpeed/blob/d24629f4fdaaa92df068de24f926d341f129112c/accelerator/real_accelerator.py#L80)). This means that it will try to use `intel_extension_for_pytorch`, which requires `oneCCL` to be compatible with `DeepSpeed`.
> 
> I could potentially include `oneCCL` as part of the environment, but I think the best path for now is to set the accelerator to cuda in the test script (which I will go ahead a do shortly). We'll leave `oneCLL` installation for a user to manually do since it isn't a requirement for most `intel_extension_for_pytorch` optimizations and then open a discussion for how/if we want to include it prebuilt in the environment itself.

Makes sense to me. @savitamittal1 what do you think?

kta-intel commented:
I add `" && export DS_ACCELERATOR='cuda'"\` to [pytorch2_0_sample_test.py](https://github.com/Azure/azureml-assets/blob/main/assets/training/general/environments/acpt-pytorch-2.0-cuda11.7/tests/pytorch2_0_sample_test.py)

Now deep speed will use cuda accelerator for the GPU workload. Hopefully this'll fix all the validation fails

Additionally, I fixed the `requirements.txt` because I noticed the updated requirements file was slightly different. Hopefully this will reflect the changes I made (just one additional line for `intel_extension_for_pytorch`) and then I added `intel_extension_for_pytorch` and `oneDNN` tags to the updated `pytorch2.0` and `tf2.12 envs`

Hopefully all merge conflicts should be resolved now and all validations pass

kta-intel commented:
@elesel I think there is an issue with the gpu-cluster `Standard_NC6` created in [create_azure_resources_daily.sh](https://github.com/Azure/azureml-assets/blob/e2d47987e0154c787ff19f68d65ddca6b2590fab/scripts/setup/create_azure_resources_daily.sh#L49) which is causing the tensorflow 2.12 test to fail.  This should run fine since I didn't change anything in the test script or the env. I could try running it on the v100 cluster if needed.

Btw, is there a Dv4-series cluster available on the test subscription? I originally planned to include a test on an Intel Xeon since the added IPEX installation is specifically optimized for it. Plus, the current cpu-cluster `Standard_DS3_v2` does not seem to have enough disk space to run the environment and tests, resulting in the following error:
```
AzureMLCompute job failed
UserScriptFilledDisk: User script filled the disk. Consider using Virtual Machine SKU with larger disk size. If the issue persists contact Azure Support.
```
However, the existing gpu tests on the updated acpt-pytorch-2.0-cuda11.7 env did pass once I set the `DS_accelerator` variable to cuda, if that is a sufficient validation

elesel commented:
> @elesel I think there is an issue with the gpu-cluster `Standard_NC6` created in [create_azure_resources_daily.sh](https://github.com/Azure/azureml-assets/blob/e2d47987e0154c787ff19f68d65ddca6b2590fab/scripts/setup/create_azure_resources_daily.sh#L49) which is causing the tensorflow 2.12 test to fail. This should run fine since I didn't change anything in the test script or the env. I could try running it on the v100 cluster if needed.
> 
> Btw, is there a Dv4-series cluster available on the test subscription? I originally planned to include a test on an Intel Xeon since the added IPEX installation is specifically optimized for it. I ran the test locally on a Standard_D4d_v4 successfully, but when I tried to push to this PR, the test failed saying that the quota limit was already reached. Alternatively, is there a larger Dv2 cluster? The `Standard_DS3_v2` does not seem to have enough disk space to run the env and test:
> 
> ```
> AzureMLCompute job failed
> UserScriptFilledDisk: User script filled the disk. Consider using Virtual Machine SKU with larger disk size. If the issue persists contact Azure Support.
> ```
> 
> However, the existing gpu tests on the updated acpt-pytorch-2.0-cuda11.7 env did pass once I set the `DS_accelerator` variable to cuda. If that is a sufficient enough validation, then we won't have to worry about running it on a cpu cluster

@kta-intel, would you please give it a shot on the V100 cluster? Unfortunately we don't have any other GPU options in our test subscription right now.

kta-intel commented:
> > @elesel I think there is an issue with the gpu-cluster `Standard_NC6` created in [create_azure_resources_daily.sh](https://github.com/Azure/azureml-assets/blob/e2d47987e0154c787ff19f68d65ddca6b2590fab/scripts/setup/create_azure_resources_daily.sh#L49) which is causing the tensorflow 2.12 test to fail. This should run fine since I didn't change anything in the test script or the env. I could try running it on the v100 cluster if needed.
> > Btw, is there a Dv4-series cluster available on the test subscription? I originally planned to include a test on an Intel Xeon since the added IPEX installation is specifically optimized for it. I ran the test locally on a Standard_D4d_v4 successfully, but when I tried to push to this PR, the test failed saying that the quota limit was already reached. Alternatively, is there a larger Dv2 cluster? The `Standard_DS3_v2` does not seem to have enough disk space to run the env and test:
> > ```
> > AzureMLCompute job failed
> > UserScriptFilledDisk: User script filled the disk. Consider using Virtual Machine SKU with larger disk size. If the issue persists contact Azure Support.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > However, the existing gpu tests on the updated acpt-pytorch-2.0-cuda11.7 env did pass once I set the `DS_accelerator` variable to cuda. If that is a sufficient enough validation, then we won't have to worry about running it on a cpu cluster
> 
> @kta-intel, would you please give it a shot on the V100 cluster? Unfortunately we don't have any other GPU options in our test subscription right now.

Thanks @elesel it looks like the V100 cluster worked

elesel commented:
> Hi @elesel. Hope you're well. I just wanted to follow up. Are we okay to merge or is there something else you need from us and/or a specific approval we're waiting on?

@kta-intel, I'm waiting on the team that owns these environments to review and take the next steps. Sorry for the delay!

skasturi commented:
@savitamittal1 @rtanase Could you please pull in relevant folks for ACPT?

savitamittal1 commented on assets/training/general/environments/acpt-pytorch-2.0-cuda11.7/tests/pytorch2_0_sample_test_ipex.py:
```diff
@@ -0,0 +1,107 @@
+    job = command(
+        code=this_dir / JOB_SOURCE_CODE,  # local path where the code is stored
+        command="pip install -r requirements.txt" \
+                " && pip uninstall accelerate -y" \
```
are there any restrictions with accelerate packagfe that we have to uninstall?

kta-intel replied:
Trying to remember back, but I recall this was actually an extension of my other reply regarding `DeepSpeed` and `intel_extension_for_pytorch`. The accelerate package seemed to have some sort of DS integration. `intel_extension_for_pytorch` needs `oneCCL` installed to be compatible. Since the installation of `oneCCL` into the curated environment wasn't in the scope of our discussions, I opted to not use accelerate instead and let a user decided if they wanted to install `oneCCL` for distributed computation capabilities. Let me know if you think otherwise (i.e. if I should just add `oneCCL`).

kta-intel replied:
Just clarify, `oneCCL` is not required by `Intel_extension_for_pytorch` in general. It is only required for distributed training/inference workloads, which is why it is not default packaged with `intel_extension_for_pytorch`

kta-intel commented on assets/training/general/environments/acpt-pytorch-2.0-cuda11.7/tests/pytorch2_0_sample_test.py:
```diff
@@ -41,6 +41,7 @@ def test_pytorch_2_0():
     job = command(
         code=this_dir / JOB_SOURCE_CODE,  # local path where the code is stored
         command="pip install -r requirements.txt" \
+                " && export DS_ACCELERATOR='cuda'"\
```
When the accelerator is not explicitly set, `DeepSpeed` will cycle through different accelerators. It will do so by checking whether certain packages are installed (`intel_extension_for_deepspeed`, `intel_extension_for_pytorch`, `torch_npu`, `torch.mps`) and setting the accelerator that corresponds to the package (`xpu`, `cpu`, `npu`, `mps`). Only after it cycles through all of these will it set `cuda`. This is not a problem if none of those packages are installed, but since we installed `intel_extension_for_pytorch`, it will try to set the accelerator to `cpu`. This was problematic for the `gpu` smoke test since it wouldn't use the correct hardware. See [real_accelerator.py](https://github.com/microsoft/DeepSpeed/blob/master/accelerator/real_accelerator.py#L108) for reference.

Setting `DS_ACCELERATOR='cuda'` forces DeepSpeed to use the `cuda` and not `cpu`

savitamittal1 replied:
So any customer using acpt image has to explicitly set ds_accelerator to cuda?

savitamittal1 replied:
As discussed offline, I would not recommend checkin this PR unless we have Deepspeed default to GPU and only enable IPEX when asked explicitly.

kta-intel replied:
listing as resolved as they are outdated. will follow up on this issue in comment chain

savitamittal1 reviewed (changes requested):
Kevin, As discussed offline, with your PR, by default ds_accelerator will be set to CPU which won’t work for customers trying to run deep learning training on GPUs. 

I wouldn’t recommend checking in this PR unless we change this behavior to have Cuda by default.

kta-intel commented:
> Kevin, As discussed offline, with your PR, by default ds_accelerator will be set to CPU which won’t work for customers trying to run deep learning training on GPUs.
> 
> I wouldn’t recommend checking in this PR unless we change this behavior to have Cuda by default.

Hi Savita,

I propose that we set the ds_accelerator env variable by default to cuda in the [dockerfile](https://github.com/kta-intel/azureml-assets/blob/kta-intel-dl-opt/assets/training/general/environments/acpt-pytorch-2.0-cuda11.7/context/Dockerfile#L16). In this way, a customer will not need to set the variable to run deep learning training on gpu. It will be set already as the environment intended and the intel extension for pytorch won't interfere.

I also removed the uninstallation of accelerator in the intel extension for pytorch test script. Seemed not to raise an issue since deepspeed is not called in the respective smoke test.

How does this sound?

savitamittal1 commented on assets/training/general/environments/acpt-pytorch-2.0-cuda11.7/tests/pytorch2_0_sample_test.py:
```diff
@@ -12,7 +12,7 @@
 BUILD_CONTEXT = Path("../context")
 JOB_SOURCE_CODE = "../../acpt-tests/src"
-TIMEOUT_MINUTES = os.environ.get("timeout_minutes", 60)
+TIMEOUT_MINUTES = os.environ.get("timeout_minutes", 720)
```
do we know why we need to increase timeout from 60 to 720?

kta-intel replied:
Oh, it timed out at 60 min on some runs for some reason, even with just the existing tests that I did not otherwise modify. I put it to 720 to avoid any timeout. I can set it back to 60 if that's preferred

kta-intel replied:
reverted to 60 min as per our offline discussion

savitamittal1 commented on assets/training/general/environments/acpt-pytorch-2.0-cuda11.7/context/Dockerfile:
```diff
@@ -13,6 +13,7 @@ RUN /var/requirements/install_system_requirements.sh && \
     rm -f /etc/nginx/sites-enabled/default
 ENV SVDIR=/var/runit
 ENV WORKER_TIMEOUT=400
+ENV DS_ACCELERATOR=cuda
```
Can you confirm if this is no longer the issue? https://github.com/microsoft/DeepSpeed/issues/3680

kta-intel replied:
Sorry for the late reply. Confirmed not to be causing an issue. Also, reading through the thread, it sounds like the print statement would've been present regardless of whether or not we set the env variable since `ds_accelerator` will be auto-detected. We explicitly set `cuda` as an env variable because it would've been auto-detected as `cpu` otherwise

kta-intel replied:
@savitamittal1 Hi, any updates?
